src/services/pallets/PalletsForeignAssetsService.ts

```diff
@@ -0,0 +1,90 @@
+// Copyright 2017-2022 Parity Technologies (UK) Ltd.
+// This file is part of Substrate API Sidecar.
+//
+// Substrate API Sidecar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import { ApiPromise } from '@polkadot/api';
+import { BlockHash } from '@polkadot/types/interfaces';
+import { AnyJson } from 'src/types/polkadot-js';
+
+import { IForeignAssetInfo, IForeignAssets } from '../../types/responses';
+import { AbstractService } from '../AbstractService';
+
+export class PalletsForeignAssetsService extends AbstractService {
+	constructor(api: ApiPromise) {
+		super(api);
+	}
+
+	/**
+	 * Fetch all foreign asset's `AssetDetails` and `AssetMetadata`.
+	 *
+	 * @param hash `BlockHash` to make call at
+	 */
+	async fetchForeignAssets(hash: BlockHash): Promise<IForeignAssets> {
+		const { api } = this;
+
+		const [{ number }, foreignAssetInfo] = await Promise.all([
+			api.rpc.chain.getHeader(hash),
+			api.query.foreignAssets.asset.entries(),
```

Comment:
Similar to below we need to correctly type the entries method. This is because both `assetStorageKeyData`, and `assetInfo` will have there own specific types.

`assetStorageKeyData: StorageKey<AnyTuple>`
`assetInfo: Option<PalletAssetsAssetDetails>`

```typescript
// Incorrect:
api.query.foreignAssets.asset.entries();

// Correct:
api.query.foreignAssets.asset.entries<Option<PalletAssetsAssetDetails>[]>()

// This is a little tricky but the above will map out to Promise<[StorageKey<AnyTuple>, Option<PalletAssetsAssetDetails>[]][]> once we input the correct type.
```

This will all ensure that the above is typed correctly so we wont need to type cast anymore and we can clean up all the types within the service.

Reply:
That is amazing!!! Thank you so much @TarikGul for spending the time to explain and showcase this. I was _fighting_ with the types quite a bit and now I get the logic! 🙏 🙏 🙏